database.cc

```diff
@@ -2326,6 +2326,68 @@ future<> database::drain() {
     co_await _commitlog->shutdown();
 }
 
+static sstring extract_cf_name(const sstring& directory_name) {
+    // cf directory is of the form: 'cf_name-timestamp'
+    return directory_name.substr(0, directory_name.find("-"));
+}
+
+future<std::vector<database::snapshot_details_result>>  database::get_snapshot_details() {
+    std::vector<sstring> data_dirs = _cfg.data_file_directories();
+    auto dirs_only_entries = lister::dir_entry_types{directory_entry_type::directory};
+    std::vector<database::snapshot_details_result> details;
+
+    for (auto& datadir : data_dirs) {
+        co_await lister::scan_dir(datadir, dirs_only_entries, [this, &dirs_only_entries, &details] (fs::path parent_dir, directory_entry de) -> future<> {
+            // KS directory
+            sstring ks_name = de.name;
+
+            co_return co_await lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, ks_name = std::move(ks_name)] (fs::path parent_dir, directory_entry de) -> future<> {
+                // CF directory
+                sstring cf_name = extract_cf_name(de.name);
+                fs::path cf_dir = parent_dir / de.name;
+
+                co_return co_await lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &dirs_only_entries, &details, &ks_name, cf_name = std::move(cf_name), cf_dir = std::move(cf_dir)] (fs::path parent_dir, directory_entry de) mutable -> future<> {
+                    // "snapshots" directory
+
+                    co_return co_await lister::scan_dir(parent_dir / de.name, dirs_only_entries, [this, &details, &ks_name, &cf_name, cf_dir] (fs::path parent_dir, directory_entry de) -> future<> {
+                        details.push_back({ 
+                            .snapshot_name = de.name,
+                            .details = {0, 0, cf_name, ks_name}
+                        });
+
+                        co_return co_await lister::scan_dir(parent_dir / de.name,  { directory_entry_type::regular }, [this, cf_dir, &details] (fs::path snapshot_dir, directory_entry de) -> future<> {
+                            auto size = co_await io_check(file_size, (snapshot_dir / de.name).native());
+
+                            // The manifest is the only file expected to be in this directory not belonging to the SSTable.
+                            // For it, we account the total size, but zero it for the true size calculation.
+                            //
+                            // All the others should just generate an exception: there is something wrong, so don't blindly
+                            // add it to the size.
+                            if (de.name != "manifest.json" && de.name != "schema.cql") {
+                                details.back().details.total += size;
```

Comment:
I only hope `scan_dir` doesn't scan the files in parallel, otherwise `details.back()` may no longer refer to what we want.

Reply:
>  I've switched to index-based access to be safe for sure

You're still using `details.back()` in one place.

BTW. Why not do the `push_back` after the `scan_dir`?

Reply:
switched to doing just that